### PR TITLE
Improve documentation for std.typecons.Proxy

### PR DESCRIPTION
**Pull request by MetaLang**
- Add a better opening explanation
- Add Params section
- Add documented unittest examples

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -4399,7 +4399,19 @@ unittest
 }
 
 /**
-Make proxy for $(D a).
+    Creates a proxy for the value `a` that will forward all operations
+    while disabling implicit conversions. The aliased item `a` must be 
+    an $(B lvalue). This is useful for creating a new type from the
```
Comment (Hackerpilot) [inline std/typecons.d]: Once I fix https://issues.dlang.org/show_bug.cgi?id=14328, these should be turned into links.

File: std/typecons.d [hunk 2]
```diff
@@ -4628,6 +4640,63 @@ unittest
...
+    NewIntType nit = 0;
+    nit++;
+    assert(nit == 1);
+    
+    
+    struct NewObjectType
+    {
+        //Ok, "new Object" is an lvalue.
+        //This code is a bit weird, though.
+        mixin Proxy!(new Object);
```
Comment (CyberShadow) [inline std/typecons.d]: What's the point of this example?

File: std/typecons.d [hunk 3]
```diff
@@ -4628,6 +4640,63 @@ unittest
...
+    }
+    
+    NewObjectType not;
+    assert(__traits(compiles, not.toHash()));
+}
+
+/**
+    There is one exception to the fact that the new type is not related to the
+    old type. Pseudo member functions that use 
+    $(LINK2 http://ddili.org/ders/d.en/ufcs.html, Universal Function Call Syntax)
```
Comment (CyberShadow) [inline std/typecons.d]: There's no reason to link to third-party resources when the official ones suffice.

http://dlang.org/function.html#pseudo-member

File: std/typecons.d (hunk 2 above)
Comment (MetaLang) [inline std/typecons.d]: To demonstrate that *any* lvalue can be used with Proxy, even if it doesn't make sense.

File: std/typecons.d (hunk 3 above)
Comment (MetaLang) [inline std/typecons.d]: Thanks, I was looking for that and couldn't find it.

**Comment by MetaLang** [comment]
I have changed the link explaining UFCS to http://dlang.org/function.html#pseudo-member and changed the documented unittest to use actual useful code.

**Comment by MetaLang** [comment]
Ping

File: std/typecons.d [hunk 4]
```diff
@@ -4628,6 +4640,65 @@ unittest
...
+        int n;
+        mixin Proxy!n;
+        
+        this(int n) { this.n = n; }
+    }
+    
+    NewIntType nit = 0;
+    nit++;
+    assert(nit == 1);
+
```
Comment (quickfur) [inline std/typecons.d]: What about showing that the proxied type is not implicitly convertible to/from the base value, as well as constitute a distinct type?

Comment (MetaLang) [inline std/typecons.d]: That's covered by the unit test above this one.

Comment (quickfur) [inline std/typecons.d]: Ah I see, thanks.

**Comment by quickfur** [comment]
Let's merge. Further improvements can be done in a followup PR.

**Comment by quickfur** [comment]
Auto-merge toggled on

**Pull request merged**